lib/spack/spack/database.py

```diff
@@ -339,7 +361,161 @@ def read_transaction(self):
         """Get a read lock context manager for use in a `with` block."""
         return ReadTransaction(self.lock, acquire=self._read)
 
-    def prefix_lock(self, spec):
+    def _failed_spec_path(self, spec):
```

Comment:
I would consider pulling the failure lock stuff out of `database.py` and putting it into a separate class somewhere with a high level description.  It can still be a member of `database`, but there's a lot in database already and the failure locks are almost a separate concern.  We can probably do that later, but it's worth not putting everything into database.

Reply:
Okay.